Fix TableGenerator using multiple NUMA nodes for allocation

lanice opened the pull request:
The addition in #731 that should have enabled multi-node usage on NUMA systems in the TableGenerator does not actually allocate the memory on the specified nodes, it only passes an allocator to the Chunk, which does nothing with it.
The memory usage of a table generation of 4000 MBs (100 Chunks * 10.000.000 Rows * sizeof(int) ) looks like this in the current master:
```
Per-node process memory usage (in MBs) for PID 24719 (hyrisePlayground)
                           Node 0          Node 1          Node 2          Node 3           Total
                  --------------- --------------- --------------- --------------- ---------------
Huge                         0.00            0.00            0.00            0.00            0.00
Heap                         0.59            0.00            0.04            3.71            4.34
Stack                        0.01            0.00            0.01            0.01            0.03
Private                    600.57            3.70           57.85         5233.83         5895.95
----------------  --------------- --------------- --------------- --------------- ---------------
Total                      601.16            3.70           57.90         5237.55         5900.31
```
With this PR, it looks more like this:
```
Per-node process memory usage (in MBs) for PID 4020 (hyrisePlayground)
                           Node 0          Node 1          Node 2          Node 3           Total
                  --------------- --------------- --------------- --------------- ---------------
Huge                         0.00            0.00            0.00            0.00            0.00
Heap                         0.00            0.05            0.61            0.00            0.66
Stack                        0.00            0.00            0.02            0.00            0.03
Private                    825.32          483.46         4056.35          541.03         5906.16
----------------  --------------- --------------- --------------- --------------- ---------------
Total                      825.32          483.51         4056.99          541.03         5906.85
```

The memory is *better* divided among the 4 nodes. As you can see there is a mysterious memory overhead, since 4000 MBs of data, AFTER compression, result in ~5900 MBs of memory usage. I am investigating this at the moment, and it is very likely that after/during the DictionaryCompression, the not-used-anymore ValueColumns do not get cleaned up correctly. This is not part of this PR though.

This also moves the TableGenerator from benchmark/ to benchmarklib/, since I think that makes sense and it makes it accessible from e.g. the playground that just links agains benchmarklib with no cost.

lanice commented on src/benchmarklib/table_generator.cpp:
```diff
@@ -108,8 +107,30 @@ std::shared_ptr<Table> TableGenerator::generate_table(
+  auto allocator_ptr_base_column = PolymorphicAllocator<std::shared_ptr<BaseColumn>>{};
+  auto allocator_value_column_int = PolymorphicAllocator<ValueColumn<int>>{};
+  auto allocator_chunk = PolymorphicAllocator<Chunk>{};
+  auto allocator_int = PolymorphicAllocator<int>{};
```
For anyone who looks at this: I am not 100% sure if I am doing this correctly. Having to use the proper types `PolymorphicAllocator<BaseColumn/Chunk/int/etc>` in the specific cases (see lines 133, 180, 181, 186) makes sense here, right?

mrzzzrm replied:
Not sure how this would work. The data is in `value_vectors` and those vector just use the default allocator, and later you just move the data out of them into the ValueColumns and that should just take the pointer, not copy any data to nodes. 
Just a quick look from my side, maybe I miss something.

lanice replied:
That is correct. For ValueColumns, it is currently *not possible* to specify the Node they are placed on (see #402 for discussion). This will be changed once [this](https://github.com/hyrise/hyrise/issues/281#issuecomment-337953032) happens.

The placement on the desired Node happens in the DictionaryCompression, because for DictionaryColumns it *is possible*. Therefore we do need to set the correct Allocators in the ValueColumn, because although the ValueColumn does not use them itself, the DictionaryCompression fetches the Allocators from the ValueColumn for correct placement.

lanice replied:
This also answers your comment below, because yes, at this point (ValueColumn generation) we don't actually place the data on the correct nodes.

mrzzzrm replied:
Alright, gottcha. Pls add a comment explaining this somewhere in the method if its not explained already and I overlooked it.

lanice replied:
Yeah good point, I will add a comment in the header explaining that `use_multiple_numa_nodes` can only be used with compression because of bla bla

mrks replied:
About the original question: Yes and no. As far as I understand it, the template parameter is just so that the returned memory has the right type. The algorithm only cares for the size.

lanice replied:
Sounds about right. But in both cases, type and size, the allocator should match the type it is used for, or not?

mrzzzrm commented:
Github won't let me comment on those lines :angry: So here `value_vectors.emplace_back(tbb::concurrent_vector<int>(chunk_size));`, line 96 the vectors are created without any allocator, that means all the generated data will just be placed wherever the default allocator puts it. Or doesn't it?

lanice commented:
With #731 being reverted, this PR still adds the functionality of distributing (dict encoded) chunks on multiple NUMA nodes during table generation, which I need and is probably good to have anyway.

I updated to the new master and added the comment regarding the allocation/placement and compression.

mrks commented:
> the not-used-anymore ValueColumns do not get cleaned up correctly.

Many allocators do not give memory back to the OS even if it is unused. In some cases, they can't, for example when a page is allocated and all but one byte are freed up again.

mrks commented on src/benchmarklib/table_generator.cpp:
```diff
@@ -105,8 +107,30 @@ std::shared_ptr<Table> TableGenerator::generate_table(
   for (ChunkID chunk_index{0}; chunk_index < num_chunks; ++chunk_index) {
-    ChunkColumns columns;
+#if HYRISE_NUMA_SUPPORT
+    if (numa_distribute_chunks) {
```
Is there any case in which we don't want the data to be distributed on a NUMA system?

I'm a bit unhappy with `numa_distribute_chunks` doing nothing if we don't have numa support. Failing is not a good option either.

lanice replied:
Hmm, I don't know. Since this is for benchmarking, I could see that you would want the option to have both, distributed, and not.

mrks replied:
Ok, but even then you should not be able to explicitly set an option that gets ignored in some cases. If `false` is the default, we could add a check that numa support has to be enabled for `true`.

mrks replied:
Not sure. The safe way is to throw an exception and wait for someone to complain about it.

mrks commented:
I'm not sure how much I like the notion of having a "next" memory resource, because it implies some ordering. I would have called it "any" memory resource.

Since the whole question of where to place new things is largely WIP, we don't have to decide on this now but can see, where it brings us.
